(un)reachable functions: valid JSON even without source locations

tautschnig opened the pull request:
We may lack source locations for generated functions such as C atomics,
and must not generate invalid JSON in such cases. Previously we would
generate output such as

```
{
  "file name": "/",
  "first line": ,
  "function": "__atomic_store_*{V}$V$",
  "last line": 0
}
```

which will now be

```
{
  "file name": "",
  "first line": 0,
  "function": "__atomic_store_*{V}$V$",
  "last line": 0
},
```

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
(un)reachable functions: valid JSON even without source locations

We may lack source locations for generated functions such as C atomics,
and must not generate invalid JSON in such cases. Previously we would
generate output such as

```
{
  "file name": "/",
  "first line": ,
  "function": "__atomic_store_*{V}$V$",
  "last line": 0
}
```

which will now be

```
{
  "file name": "",
  "first line": 0,
  "function": "__atomic_store_*{V}$V$",
  "last line": 0
},
```

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 963423de500bd198e54335777f2664df75139bf5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118205389

peterschrammel commented on src/goto-analyzer/unreachable_instructions.cpp:
```diff
@@ -251,12 +267,9 @@ static void json_output_function(
-       id2string(first_location.get_file())))},
-    {"first line", json_numbert(id2string(first_location.get_line()))},
-    {"last line", json_numbert(id2string(last_location.get_line()))}};
+    {"file name", json_stringt{file_name_string(first_location)}},
```
I'm not sure who the user of this is, but the format doesn't match Google JSON style that we use throughout the JSON output (properties use camelCase). It's an opportunity to fix that now.

tautschnig replied:
I've noticed the same when fixing this. One user is CBMC viewer, see #4863. We would just need to make sure we fix both of them in a synchronised way.

peterschrammel commented on src/goto-analyzer/unreachable_instructions.cpp:
```diff
@@ -270,12 +283,9 @@ static void xml_output_function(
-                    id2string(first_location.get_file())));
-  x.set_attribute("first line", id2string(first_location.get_line()));
-  x.set_attribute("last line", id2string(last_location.get_line()));
+  x.set_attribute("file name", file_name_string(first_location));
```
```suggestion
  x.set_attribute("file_name", file_name_string(first_location));
```
etc

tautschnig committed:
(un)reachable functions: normalize JSON and XML keys

JSON uses camelCase, XML uses underscores or dashes (we liberally mix
them across the code base).

tautschnig commented:
@peterschrammel Could you please take another look - I've added a second commit addressing the key naming.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: cb2f8de7c7e53c2e8c333d4114f513abfda361c3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/122266011

peterschrammel commented on src/goto-analyzer/unreachable_instructions.cpp:
```diff
@@ -243,22 +250,26 @@ bool static_unreachable_instructions(
-       id2string(first_location.get_file())))},
-    {"first line", json_numbert(id2string(first_location.get_line()))},
-    {"last line", json_numbert(id2string(last_location.get_line()))}};
+  json_objectt entry{{"function", json_stringt(function)},
```
It would be preferable to omit the property if we don't have information instead of using an empty string and 0.

tautschnig replied:
Good call, that's now done.
